server/irmaserver/sessions.go

```diff
@@ -138,58 +156,194 @@ func (s *memorySessionStore) deleteExpired() {
 
 	// Using a write lock, delete the expired sessions
 	s.Lock()
-	for _, token := range expired {
-		session := s.requestor[token]
+	for _, Token := range expired {
+		session := s.requestor[Token]
 		if session.sse != nil {
-			session.sse.CloseChannel("session/" + session.token)
-			session.sse.CloseChannel("session/" + session.clientToken)
+			session.sse.CloseChannel("session/" + session.Token)
+			session.sse.CloseChannel("session/" + session.ClientToken)
 		}
-		delete(s.client, session.clientToken)
-		delete(s.requestor, token)
+		delete(s.client, session.ClientToken)
+		delete(s.requestor, Token)
 	}
 	s.Unlock()
 }
 
+// MarshalJSON marshals sessionData to be used in the Redis in-memory datastore.
+func (s *sessionData) MarshalJSON() ([]byte, error) {
+	return json.Marshal(*s)
+}
+
+// UnmarshalJSON unmarshals sessionData.
+func (s *sessionData) UnmarshalJSON(data []byte) error {
+	type rawSessionData sessionData
+
+	var temp struct {
+		Rrequest *json.RawMessage `json:",omitempty"`
+		rawSessionData
+	}
+
+	if err := json.Unmarshal(data, &temp); err != nil {
+		return err
+	}
+
+	*s = sessionData(temp.rawSessionData)
+
+	if temp.Rrequest == nil {
+		s.Rrequest = nil
+		return errors.Errorf("temp.Rrequest == nil: %d \n", temp.Rrequest)
+	}
+
+	// unmarshal Rrequest
+	ipR := &irma.IdentityProviderRequest{}
+	spR := &irma.ServiceProviderRequest{}
+	sigR := &irma.SignatureRequestorRequest{}
+
+	if err := json.Unmarshal(*temp.Rrequest, ipR); err == nil && s.Action == "issuing" {
+		s.Rrequest = ipR
+	} else if err = json.Unmarshal(*temp.Rrequest, spR); err == nil && s.Action == "disclosing" {
+		s.Rrequest = spR
+	} else if err = json.Unmarshal(*temp.Rrequest, sigR); err == nil && s.Action == "signing" {
+		s.Rrequest = sigR
+	} else {
+		return err
+	}
+
+	return nil
+}
+
+func (s *redisSessionStore) get(t string) (*session, error) {
+	//TODO: input validation string?
+	val, err := s.client.Get(context.Background(), tokenLookupPrefix+t).Result()
+	if err == redis.Nil {
+		return nil, nil
+	} else if err != nil {
+		_ = server.LogError(err)
+		return nil, errors.New("redis error")
```

Comment:
You might want to use `errors.WrapError(...)`. Then the original error is also included, which might be nice when people send us stack traces for debugging.

(This also holds for all other places where `errors.New("redis error")` is thrown)

Reply:
You already solved this my introducing a custom error right?

Reply:
Yep